src/backend/optimizer/plan/createplan.c

```diff
@@ -6471,6 +6475,132 @@ make_modifytable(PlannerInfo *root,
 	return node;
 }
 
+/*
+ * can_elide_explicit_motion_recurse
+ *	  Recursive guts of can_elide_explicit_motion().
+ *
+ *	  Check for motions by recursively looking at right subtree first. If it does
+ *	  not exist, go into left subtree. If we find a motion before the scan, we
+ *	  return false and exit from that subtree. A scan is always underneath the
+ *	  motion, so we won't encounter the scan if we exited the subtree. In that
+ *	  case, we can't elide Explicit Redistribute Motion.
+ *
+ *	  If we encounter the scan before any motions, then we can elide
+ *	  unneccessary Explicit Redistribute Motion.
+ */
+static bool
+can_elide_explicit_motion_recurse(Plan *plan, List *rtable, Oid relid,
+								  bool relid_is_subclass, Oid reltypeid)
+{
+	while (plan)
```

Comment:
Why not use a walker?

Reply:
What walker? Should this function be moved somewhere else?

Reply:
plan_tree_walker

Reply:
`plan_tree_walker` looks into left tree first, but as I have noticed during testing, scans and motions are usually in the right subtree. So, recursing into right subtree first can save time. Also, it doesn't need to look at any of the node members except for targetlist and scanrelid.

Reply:
Can you explain why the right subrtree is more frequent? It should be explained in the commit description. Why should only targetlist and scanrelid be visited?

Reply:
We're searching for relation ID that is modified in the `ModifyTable` from `DELETE`/`UPDATE`. 

For motions, checking the target list is needed to see if motion affects any of columns that are from relation we're searching for. Their member `resno` contains index into the range table, using which we can retrieve the relation ID. 

For scans (which all should be able to be cast to `Scan` structure) `scanrelid` member of `Scan` contains the index into the range table for scanned relation.

We need only these items, because our main goal is to check affected relations of motions and scans and find out what comes first. Specifically if we look at the walker function, we don't need to check any conditions, `InitPlans` [are already handled](https://github.com/arenadata/gpdb/blob/ee5cbe6d1c276c77342ff749e42f7d3c573d8598/src/backend/cdb/cdbmutate.c#L756-L761), and `flow` nodes do not contain useful information in this case.

Assumption that the right subrtree is more frequent comes from my observation while in looking at this patch in a debugger. A scan or motion was usually found in second or third recursion into `can_elide_explicit_motion_recurse`. Below are found examples from new tests.

Scan on relevant table is in right subtree:
```
                         QUERY PLAN                         
------------------------------------------------------------
 Update on t1
   ->  Hash Join
         Hash Cond: (t_strewn.i = t1.i)
         ->  Redistribute Motion 3:3  (slice1; segments: 3)
               Hash Key: t_strewn.i
               ->  Seq Scan on t_strewn
         ->  Hash
               ->  Seq Scan on t1
```

Scan is in the right subtree again:
```
                                         QUERY PLAN                                         
--------------------------------------------------------------------------------------------
 Delete on public.foo
   ->  Explicit Redistribute Motion 3:3  (slice2; segments: 3)
         Output: foo.ctid, foo.gp_segment_id, i.ctid
         ->  Hash Join
               Output: foo.ctid, foo.gp_segment_id, i.ctid
               Hash Cond: (i.j = foo.f1)
               ->  Seq Scan on public.i
                     Output: i.ctid, i.j
               ->  Hash
                     Output: foo.ctid, foo.gp_segment_id, foo.f1
                     ->  Broadcast Motion 3:3  (slice1; segments: 3)
                           Output: foo.ctid, foo.gp_segment_id, foo.f1
                           ->  Seq Scan on public.foo
                                 Output: foo.ctid, foo.gp_segment_id, foo.f1
```

While deleting in partitions, the relevant scan is in the still right subtree:
```
                                                QUERY PLAN                                                 
-----------------------------------------------------------------------------------------------------------
 Delete on public.t1_1_prt_1
   ->  Explicit Redistribute Motion 3:3  (slice2; segments: 3)
         Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t2_1_prt_def.ctid, t2_1_prt_def.tableoid
         ->  Hash Join
               Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t2_1_prt_def.ctid, t2_1_prt_def.tableoid
               Hash Cond: (t2_1_prt_def.a = t1_1_prt_1.a)
               ->  Append
                     ->  Result
                           Output: t2_1_prt_def.ctid, t2_1_prt_def.tableoid, t2_1_prt_def.a
                           One-Time Filter: PartSelected
                           ->  Seq Scan on public.t2_1_prt_def
                                 Output: t2_1_prt_def.ctid, t2_1_prt_def.tableoid, t2_1_prt_def.a
                     ->  Result
                           Output: t2_1_prt_2.ctid, t2_1_prt_2.tableoid, t2_1_prt_2.a
                           One-Time Filter: PartSelected
                           ->  Seq Scan on public.t2_1_prt_2
                                 Output: t2_1_prt_2.ctid, t2_1_prt_2.tableoid, t2_1_prt_2.a
                     ->  Result
                           Output: t2_1_prt_3.ctid, t2_1_prt_3.tableoid, t2_1_prt_3.a
                           One-Time Filter: PartSelected
                           ->  Seq Scan on public.t2_1_prt_3
                                 Output: t2_1_prt_3.ctid, t2_1_prt_3.tableoid, t2_1_prt_3.a
                     ->  Result
                           Output: t2_1_prt_4.ctid, t2_1_prt_4.tableoid, t2_1_prt_4.a
                           One-Time Filter: PartSelected
                           ->  Seq Scan on public.t2_1_prt_4
                                 Output: t2_1_prt_4.ctid, t2_1_prt_4.tableoid, t2_1_prt_4.a
                     ->  Result
                           Output: t2_1_prt_5.ctid, t2_1_prt_5.tableoid, t2_1_prt_5.a
                           One-Time Filter: PartSelected
                           ->  Seq Scan on public.t2_1_prt_5
                                 Output: t2_1_prt_5.ctid, t2_1_prt_5.tableoid, t2_1_prt_5.a
               ->  Hash
                     Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t1_1_prt_1.a
                     ->  Partition Selector for t2 (dynamic scan id: 2)
                           Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t1_1_prt_1.a
                           Filter: t1_1_prt_1.a
                           ->  Broadcast Motion 3:3  (slice1; segments: 3)
                                 Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t1_1_prt_1.a
                                 ->  Seq Scan on public.t1_1_prt_1
                                       Output: t1_1_prt_1.ctid, t1_1_prt_1.gp_segment_id, t1_1_prt_1.a
```

Reply:
> Assumption that the right subrtree is more frequent comes from my observation while in looking at this patch in a debugger. 

You shouldn't use such motivation when make decision. For example, relation size makes sense in your 1st example. Moreover, test examples usually have little in common with production workload.